x/subscription/keeper/subscription_test.go

```diff
@@ -0,0 +1,187 @@
+	items := make([]types.Subscription, n)
+	_, creator := sigs.GenerateFloatingKey()
+
+	for i := range items {
+		items[i].Creator = creator.String()
+		items[i].Consumer = "consumer-" + strconv.Itoa(i)
+		items[i].Block = ctx.BlockHeight()
+		items[i].PackageIndex = "testpkg"
+		items[i].PackageBlock = ctx.BlockHeight()
+
+		keeper.SetSubscription(ctx, items[i])
+	}
+	return items
+}
+
+func TestSubscriptionGet(t *testing.T) {
+	keeper, ctx := keepertest.SubscriptionKeeper(t)
+	items := createNSubscription(keeper, ctx, 10)
+	for _, item := range items {
+		rst, found := keeper.GetSubscription(ctx,
+			item.Consumer,
+		)
+		require.True(t, found)
+
+		require.Equal(t,
+			nullify.Fill(&item),
+			nullify.Fill(&rst),
+		)
+	}
+}
+
+func TestSubscriptionRemove(t *testing.T) {
+	keeper, ctx := keepertest.SubscriptionKeeper(t)
+	items := createNSubscription(keeper, ctx, 10)
+	for _, item := range items {
+		keeper.RemoveSubscription(ctx,
+			item.Creator,
+		)
+		_, found := keeper.GetSubscription(ctx,
+			item.Creator,
+		)
+		require.False(t, found)
+	}
+}
+
+func TestSubscriptionGetAll(t *testing.T) {
+	keeper, ctx := keepertest.SubscriptionKeeper(t)
+	items := createNSubscription(keeper, ctx, 10)
+	require.ElementsMatch(t,
+		nullify.Fill(items),
+		nullify.Fill(keeper.GetAllSubscription(ctx)),
+	)
+}
+
+func TestCreateSubscription(t *testing. T) {
+	_, keepers, _ctx := keepertest.InitAllKeepers(t)
+	ctx := sdk.UnwrapSDKContext(_ctx)
+
+	keeper := keepers.Subscription
+	bank := keepers.BankKeeper
+
+	creators := []struct {
+		address string
+		amount  int64
+	}{
+		{
+			address: "FILL",
+			amount:  100000,
+		},
+		{
+			address: "FILL",
+			amount:  1,
+		},
+		{
+			address: "invalid creator",
+			amount:  0,
+		},
+	}
+
+	for i := range creators {
+		if creators[i].address == "FILL" {
+			_, addr := sigs.GenerateFloatingKey()
+			creators[i].address = addr.String()
+			coins := sdk.NewCoins(sdk.NewCoin(
+				epochstoragetypes.TokenDenom,
+				sdk.NewInt(creators[i].amount),
+			))
+			bank.SetBalance(ctx, addr, coins)
+		}
+	}
+
+	consumers := make([]string, 4)
+	for i := range consumers {
+		_, addr := sigs.GenerateFloatingKey()
+		consumers[i] = addr.String()
+	}
+	consumers[3] = "invalid consumer"
+
+	template := []struct {
+		name      string
+		index     string
+		creator   int
+		consumers []int
+		success   bool
+	}{
+		{
+			name:      "create subscriptions",
+			index:     "testpkg",
+			creator:   0,
+			consumers: []int{0, 1},
+			success:   true,
+		},
+		{
+			name:      "invalid creator",
+			index:     "testpkg",
+			creator:   2,
+			consumers: []int{2},
+			success:   false,
+		},
+		{
+			name:      "invalid consumer",
+			index:     "testpkg",
+			creator:   0,
+			consumers: []int{3},
+			success:   false,
+		},
+		{
+			name:      "insufficient funds",
+			index:     "testpkg",
+			creator:   1,
+			consumers: []int{2},
+			success:   false,
+		},
+		{
+			name:      "invalid package",
+			index:     "",
```

Comment:
this check will fail in the ValidateBasic() function since the index is empty. Maybe add a test with a non-empty plan index that doesn't exist?

Reply:
Sure.